conda_forge_tick/git_utils.py

```diff
@@ -88,77 +97,533 @@ def github_client() -> github.Github:
+        self, git_dir: Path, branch: str, start_point: str | None = None
+    ):
+        """
+        Checkout a new branch in a git repository.
+        :param git_dir: The directory of the git repository.
+        :param branch: The name of the new branch.
+        :param start_point: The name of the branch to branch from, or None to branch from the current branch.
+        """
+        start_point_option = [start_point] if start_point else []
+
+        self._run_git_command(
+            ["checkout", "--quiet", "-b", branch] + start_point_option, git_dir
+        )
+
+    def clone_fork_and_branch(
+        self,
+        origin_url: str,
+        target_dir: Path,
+        upstream_url: str,
+        new_branch: str,
+        base_branch: str = "main",
+    ):
+        """
+        Convenience method to do the following:
+        1. Clone the repository at origin_url into target_dir (resetting the directory if it already exists).
+        2. Add a remote named "upstream" with the URL upstream_url (ignoring if it already exists).
+        3. Fetch all changes from all remotes.
+        4. Checkout the base branch.
+        5. Create a new branch from the base branch with the name new_branch.
+
+        This is usually used to create a new branch for a pull request. In this case, origin_url is the URL of the
+        user's fork, and upstream_url is the URL of the upstream repository.
+
+        :param origin_url: The URL of the repository (fork) to clone.
+        :param target_dir: The directory to clone the repository into.
+        :param upstream_url: The URL of the upstream repository.
+        :param new_branch: The name of the branch to create.
+        :param base_branch: The name of the base branch to branch from.
+
+        :raises GitCliError: If a git command fails.
+        """
+        self.clone_repo(origin_url, target_dir)
+
+        try:
+            self.add_remote(target_dir, "upstream", upstream_url)
+        except GitCliError as e:
+            logger.debug(
+                "It looks like remote 'upstream' already exists. Ignoring.", exc_info=e
+            )
+            pass
+
+        self.fetch_all(target_dir)
+
+        if self.does_branch_exist(target_dir, base_branch):
+            self.checkout_branch(target_dir, base_branch)
+        else:
+            try:
+                self.checkout_branch(target_dir, f"upstream/{base_branch}", track=True)
+            except GitCliError as e:
+                logger.debug(
+                    "Could not check out with git checkout --track. Trying git checkout -b.",
+                    exc_info=e,
+                )
+
+                # not sure why this is needed, but it was in the original code
+                self.checkout_new_branch(
+                    target_dir,
+                    base_branch,
+                    start_point=f"upstream/{base_branch}",
+                )
+
+        # not sure why this is needed, but it was in the original code
```

Comment:
Thanks for leaving this as is! The code here has years of weird bugs and edge cases. Not all of them were documented. :/